### PR TITLE
`core::error`, `no_std`

### PR DESCRIPTION
**Pull request by jordens**
[`error_in_core`](https://github.com/rust-lang/rust/issues/103765) has been stabilized in 1.81. This PR builds on the work of #64/#211/#244.

* Introduce a `std` feature (default enabled) to gate the `std` components (currently only `std::path`)
* Private re-export of `std::error` or `core::error` to avoid bumping MSRV for `std`. `no_std` does require `1.81`.
* Tests/docs/CI updates
* Macro hygiene: impl/expand.rs uses full absolute paths.

It's unclear to me whether it's better to bump the MSRV (`1.56` -> `1.81`) or use a private re-export of either `std::error` or `core::error` (as done here) to keep support for `rustc < 1.81`.

### CODE REVIEW

**Comment by jstarks** [comment]
Given the wide use of this crate, it seems very unlikely that an MSRV bump would be accepted.

**Comment by jordens** [comment]
@dtolnay what's your take on the PR, especially regarding the MSRV question?

> It's unclear to me whether it's better to bump the MSRV (1.56 -> 1.81 as done in this draft) or use a private re-export of either std::error or core::error (as done [previously](https://github.com/dtolnay/thiserror/blob/3663fd16b7ae40c966e1b7e8926a2ca4164730be/src/lib.rs#L266-L271)) to keep support for rustc < 1.81.

**Comment by umgefahren** [comment]
What if one is just able to somehow pass `core::error` instead of `std::error`?

**Comment by kayabaNerve** [comment]
My personal belief is `default-features = false` should be MSRV 1.81 and `features = ["std"]` should be MSRV 1.56.

**Comment by jordens** [comment]
@dtolnay any guidance on how you'd like to see it done?

**Comment by kayabaNerve** [comment]
PR LGTM with edge case that I don't believe you can import thiserror with a distinct package name with this path (I'm unsure if you could already).

**Comment by jordens** [comment]
@dtolnay I know this is at least the third PR attempting it and this has come up many times in the past. But now that your [condition](https://github.com/dtolnay/thiserror/pull/211#pullrequestreview-1530568192) can be met, what's missing? And what can we do to get this in?

**Comment by ognevny** [comment]
this code doesn't work with my #![no_std] lib...
```
60 | #[derive(Debug, Error, PartialEq, Eq)]
   |                 ^^^^^
   |                 |
   |                 use of undeclared crate or module `std`
   |                 in this derive macro expansion
   |
  ::: ~/.cargo/git/checkouts/thiserror-8e8b6321d051832b/c9cd3b9/impl/src/lib.rs:33:1
   |
33 | pub fn derive_error(input: TokenStream) -> TokenStream {
   | ------------------------------------------------------ in this expansion of `#[derive(Error)]`
   |
help: consider importing this module
   |
5  + use core::error;
```

**Comment by ZhekaS** [comment]
> this code doesn't work with my #![no_std] lib...


I am getting a similar one for a bin crate:
```
error[E0433]: failed to resolve: use of undeclared crate or module `std`
  --> src/********:49:17
   |
49 | #[derive(Debug, Error)]
   |                 ^^^^^ use of undeclared crate or module `std`
   |
   = note: this error originates in the derive macro `Error` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this module
   |
1  + use core::error;
   |

For more information about this error, try `rustc --explain E0433`.
```
Looks like it only happens when `fn source` is derived. Then the macro expansion looks like:

`fn source(&self) -> ::core::option::Option<&(dyn std::error::Error + 'static)> ....`

Note the `std` there...

**Comment by jdygert-spok** [comment]
> PR LGTM with edge case that I don't believe you can import thiserror with a distinct package name with this path (I'm unsure if you could already).

Looks like you already couldn't, at least when using certain attributes, since there's existing use of the `__private` module.

> this code doesn't work with my #![no_std] lib...

Missed one: https://github.com/dtolnay/thiserror/pull/304/files#diff-bb7dbcae2958da4ff43be22e512a8abfb22393983d7e49717d5a8ec25ea7abaeR270

I think this PR would be more likely to be merged if CI would catch this.

**Comment by ZhekaS** [comment]
> Missed one: https://github.com/dtolnay/thiserror/pull/304/files#diff-bb7dbcae2958da4ff43be22e512a8abfb22393983d7e49717d5a8ec25ea7abaeR270
>
Works like a charm with the above fix on a somewhat large `no_std` project. Really need it to get in ASAP.

**Comment by jordens** [comment]
Fixed in the commit above.

**Comment by ognevny** [comment]
now it fails to publish my lib into crates-io
```
error[E0433]: failed to resolve: use of undeclared crate or module `std`
  --> src/....rs:62:17
   |
62 | #[derive(Debug, Error, PartialEq, Eq)]
   |                 ^^^^^
   |                 |
   |                 use of undeclared crate or module `std`
   |                 in this derive macro expansion
   |
  ::: /home/ogn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/thiserror-impl-1.0.63/src/lib.rs:33:1
   |
33 | pub fn derive_error(input: TokenStream) -> TokenStream {
   | ------------------------------------------------------ in this expansion of `#[derive(Error)]`
   |
help: consider importing this module
   |
5  + use core::error;
   |

error[E0433]: failed to resolve: use of undeclared crate or module `std`
  --> src/....rs:68:13
   |
68 |     #[error(transparent)]
   |             ^^^^^^^^^^^ use of undeclared crate or module `std`
   |
help: consider importing one of these items
   |
5  + use core::error::Error;
   |
5  + use core::fmt::Error;
   |
5  + use regex::Error;
   |

error[E0433]: failed to resolve: use of undeclared crate or module `std`
  --> src/....rs:9:17
   |
9  | #[derive(Debug, Error, PartialEq, Eq)]
   |                 ^^^^^
   |                 |
   |                 use of undeclared crate or module `std`
   |                 in this derive macro expansion
   |
  ::: /home/ogn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/thiserror-impl-1.0.63/src/lib.rs:33:1
   |
33 | pub fn derive_error(input: TokenStream) -> TokenStream {
   | ------------------------------------------------------ in this expansion of `#[derive(Error)]`
   |
help: consider importing this module
   |
5  + use core::error;
   |

For more information about this error, try `rustc --explain E0433`.
error: could not compile `...` (lib) due to 3 previous errors
```
while tests without default features (std disabled) compile and run successfully

**Comment by jordens** [comment]
You can't publish if you have git dependencies.

**Comment by jmwample** [comment]
It might be more palatable if this PR made `no-std` a feature such that enabling it bumps the MSRV to 1.81 instead of forcing an MSRV bump for the library. You could make `no-std` default for compilers 1.81+ and then an organic bump could be done when the library actually wants to drop support for older versions.

**Comment by kayabaNerve** [comment]
Features are intended to be additive. no-std disables some functionality so it isn't additive. std is additive, and when enabled, maintains the existing MSRV.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -38,6 +38,9 @@ jobs:
       - name: Enable nightly-only tests
         run: echo RUSTFLAGS=${RUSTFLAGS}\ --cfg=thiserror_nightly_testing >> $GITHUB_ENV
         if: matrix.rust == 'nightly'
+      - name: Test no-default-features
+        run: cargo test --all --no-default-features
```
Comment (domenukk) [inline .github/workflows/ci.yml:42]: This isn't enough to assert `no_std` compatibility as dependencies might still pull in `std`.
I would suggest adding something like:
```sh
cargo check --target aarch64-unknown-none --no-default-features
```

Comment (jordens) [reply .github/workflows/ci.yml:42]: There are no dependencies. This is the top level crate.

Comment (domenukk) [reply .github/workflows/ci.yml:42]: Perfect, they are rare.

Comment (domenukk) [reply .github/workflows/ci.yml:42]: Wait, I see
```rust
[dependencies]
proc-macro2 = "1.0.74"
quote = "1.0.35"
syn = "2.0.46"
```

Comment (jordens) [reply .github/workflows/ci.yml:42]: in a proc macro crate

**Comment by dhedey** [comment]
@jordens  - What do we think about using `build.rs` to effectively enable the `std::core` functionality on rustc < 1.81 ?

We could do this by introducing a `cargo:rustc-cfg=anyhow_no_core_error` rustc-cfg flag, in a very similar manner to dtolnay did in his equivalent PR in anyhow  https://github.com/dtolnay/anyhow/pull/383/files ?

**Comment by jordens** [comment]
I'd like to hear about the prospects of this PR first.

**Review by bushrat011899 — COMMENTED** [review]
I'd _really_ like this to land; the implementation looks good and even manages to avoid bumping MSRV as long as you include the `std` feature. Getting this merged would unblock some `no_std` support for Bevy that I'd love to get moving before the end of the year if possible.

**Review by BenjaminBrienen — COMMENTED** [review]
This is a good change which addresses all concerns and merging this would enable other crates to continue using thiserror. It's unfortunate that the ecosystem is splitting over this problem which is already solved.

**Comment by dhedey** [comment]
I certainly support this PR.

That said, and please correct me if I’m wrong, but it is my understanding that if we release this current PR, we immediately break the builds of anyone compiling with rustc < 1.81 which imports thiserror with default-features = false.

Given the widespread usage of thiserror, I would imagine such a breakage is not tenable.

I have suggested a simple tweak to this PR which I believe avoids this issue [in this comment](https://github.com/dtolnay/thiserror/pull/304#issuecomment-2371846932) by effectively always enabling an equivalent-to-std feature before the Error trait is available in core.

I assume given dtolnay was the author of the equivalent PR I linked to in anyhow, they would favour such a solution, and given that I assume the merging of any such PR is blocked on dtolnay’s approval, it seems likely we require such a change; tested with this particular import case.

But maybe I’m missing something which means thiserror should be given different treatment?

**Comment by jstarks** [comment]
> I certainly support this PR.
> 
> That said, and please correct me if I’m wrong, but it is my understanding that if we release this current PR, we immediately break the builds of anyone compiling with rustc < 1.81 which imports thiserror with default-features = false.
> 
> Given the widespread usage of thiserror, I would imagine such a breakage is not tenable.
> 
> I have suggested a simple tweak to this PR which I believe avoids this issue [in this comment](https://github.com/dtolnay/thiserror/pull/304#issuecomment-2371846932) by effectively always enabling an equivalent-to-std feature before the Error trait is available in core.
> 
> I assume given dtolnay was the author of the equivalent PR I linked to in anyhow, they would favour such a solution, and given that I assume the merging of any such PR is blocked on dtolnay’s approval, it seems likely we require such a change; tested with this particular import case.
> 
> But maybe I’m missing something which means thiserror should be given different treatment?

There are a few places where `std::path::Path` is referenced, as part of an `impl` on some trait. `Path` is not and will never be in `core`. So these references have to be gated behind a feature.

**Comment by dhedey** [comment]
> There are a few places where std::path::Path is referenced, as part of an impl on some trait. Path is not and will never be in core. So these references have to be gated behind a feature.

I'm not suggesting we remove the `std` feature - I'm just suggesting it's effectively auto-enabled pre rust 1.81, to minimize compilation/CI breaks for people using a rust version before it supports `core::error`.

The current PR I believe will break compilation for anyone who is building on `rustc < 81` and has imported `core::error` with `default-features = false`.

Instead, I suggest [like this PR on anyhow](https://github.com/dtolnay/anyhow/pull/383/files) we:
* Introduce in `build.rs`, `if rustc < 81` then `println!("cargo:rustc-cfg=thiserror_no_core_error");`
* In code, replace `feature = "std"` with `any(feature = "std", not(thiserror_no_core_error)),`

After this change, I believe the breaks are minimized to people who:
* Are on `rustc >= 81`
* Use `default-features = false`
* Rely on std features in this crate

Which I imagine is a sufficiently small crowd that this might be deemed acceptable, given the fix is to just add the `std` feature.

File: impl/src/expand.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ fn fallback(input: &DeriveInput, error: syn::Error) -> TokenStream {
 
         #[allow(unused_qualifications)]
         #[automatically_derived]
-        impl #impl_generics std::error::Error for #ty #ty_generics #where_clause
+        impl #impl_generics ::thiserror::__private::error::Error for #ty #ty_generics #where_clause
```
Comment (dhedey) [inline impl/src/expand.rs:40]: Out of interest, why did we choose this rather than hardcoding `::std` or `::core`, in the compiled output with e.g. an `#error_namespace` variable?

My slight concern is that use of `::thiserror::__private` may leak this `__private` namespace out and could confuse people debugging this. But using `::std` / `::core` directly it's more obvious what's failing.

File: impl/src/expand.rs [hunk 3]
```diff
@@ -307,12 +309,12 @@ fn impl_enum(input: Enum) -> TokenStream {
                     let self_provide = if type_is_option(backtrace_field.ty) {
                         quote! {
                             if let ::core::option::Option::Some(backtrace) = backtrace {
-                                #request.provide_ref::<std::backtrace::Backtrace>(backtrace);
+                                #request.provide_ref::<::std::backtrace::Backtrace>(backtrace);
```
Comment (dhedey) [inline impl/src/expand.rs:312]: MINOR: It would be nice if we gave a better compiler error if a user tried to specify a backtrace when we're not in `std` mode, which could advise the user to enable the `std` feature to use this?

I believe the right place for this would be in `valid.rs`.

File: Cargo.toml [hunk 4]
```diff
@@ -11,6 +11,8 @@ license = "MIT OR Apache-2.0"
 repository = "https://github.com/dtolnay/thiserror"
 rust-version = "1.56"
 
+# without the `std` feature (i.e. `no_std`): rust-version = "1.81"
```
Comment (dhedey) [inline Cargo.toml:14]: As mentioned in [this comment](https://github.com/dtolnay/thiserror/pull/304#issuecomment-2407101161), I feel we should attempt to widen our build compatibility to not break `rustc < 1.81` builds importing `thiserror` with `default-features = false`.

**Review by dhedey — COMMENTED** [review]
Just a few thoughts/suggestions from a bystander.

**Comment by dhedey** [comment]
Actually, I'm a little wrong, apologies. I think I misinterpreted [the PR on anyhow](https://github.com/dtolnay/anyhow/pull/383).

The PR on anyhow simply avoids outputting a core error at all if `anyhow_no_core_error` is true.

So, when upgrading to `anyhow = { version = "1.0.89", default-features = false }` with rustc at 1.80, it appears the crate itself compiles, but upstream crates are broken because it no longer outputs conversions from `std::error::Error` to `anyhow::Error`:
```rust
fn main() -> Result<()> {
    let config = std::fs::read_to_string("cluster.json")?;
    Ok(())
}
```
<details>
<summary>Compilation Error</summary>
<p>

```text
error[E0277]: `?` couldn't convert the error to `anyhow::Error`
 --> src/main.rs:4:57
  |
3 | fn main() -> Result<()> {
  |              ---------- expected `anyhow::Error` because of this
4 |     let config = std::fs::read_to_string("cluster.json")?;
  |                  ---------------------------------------^ the trait `From<std::io::Error>` is not implemented for `anyhow::Error`, which is required by `Result<(), anyhow::Error>: FromResidual<Result<Infallible, std::io::Error>>`
  |                  |
  |                  this can't be annotated with `?` because it has type `Result<_, std::io::Error>`
  |
  = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
  = help: the following other types implement trait `FromResidual<R>`:
            <Result<T, F> as FromResidual<Result<Infallible, E>>>
            <Result<T, F> as FromResidual<Yeet<E>>>
  = note: required for `Result<(), anyhow::Error>` to implement `FromResidual<Result<Infallible, std::io::Error>>`
```

</p>
</details>

That said, unlike anyhow, this crate (and indeed the output from the macro) simply wouldn't compile on 1.80 with `default-features = false`, so I think personally I'd still advocate for this.

But I accept it might be acceptable. Maybe I'm strange to default to using `default-features = false` for all imports and only turning on what I need.

**Comment by jdygert-spok** [comment]
To be clear, `anyhow` already had an `std` feature enabled by default, meaning that PR strictly added functionality when `core::error::Error` is available. `thiserror` does NOT currently have an `std` feature, meaning this is a breaking change, depending on how you view default features.

**Comment by kayabaNerve** [comment]
I think all of this can be argued bikeshedding and is best left to a further PR. Someone can even now prepare a PR, descending from these commits, ready to be merged after this PR is merged if it's acceptably done and desirable. It doesn't have to be a blocker here.

**Comment by domenukk** [comment]
As another reference point, rust-url - with a similar amount of users on crates.io - just accepted a PR that adds a std feature. Just as in this PR, MSRV for std stayed the same, no_std requires 1.81.
https://github.com/servo/rust-url/pull/831

TL;DR: Would be great to get this PR merged as well :)

**Comment by clabby** [comment]
Any updates here? This PR looks good to go, other popular crates have adopted this same pattern. @dtolnay, any notes on what's blocking this on your end?

**Comment by juliankrieger** [comment]
Is there any work left to do here? I would love to help.

**Comment by zdivelbiss** [comment]
I'd love to use this crate in a project of mine, and it seems there hasn't been much activity in describing what needs to be done / changed.

@dtolnay Any input on how this can proceed to being merged?

**Review by dtolnay — COMMENTED** [review]
I had a chance to work on this feature this week, and I ended up going with a different implementation that automatically enables a std dependency in compilers older than 1.81, more like my corresponding change in `anyhow` from earlier which has been linked above, and along the lines of what was suggested by @dhedey in https://github.com/dtolnay/thiserror/pull/304#issuecomment-2371846932 and https://github.com/dtolnay/thiserror/pull/304#issuecomment-2404350218. There were also some future compatibility drawbacks of this PR that I handled differently, which you can find in #373 and #372.

Thanks anyway for the work on this PR, and also to @dhedey for numerous substantive comments.
